moveit_core/trajectory_processing/src/time_optimal_trajectory_generation.cpp

```diff
@@ -0,0 +1,1012 @@
+
+      startConfig = blendSegment->getConfig(blendSegment->getLength());
+    }
+    else
+    {
+      pathSegments.push_back(new LinearPathSegment(startConfig, *config2));
+      startConfig = *config2;
+    }
+    config1 = config2;
+    config2++;
+  }
+
+  // Create list of switching point candidates, calculate total path length and
+  // absolute positions of path segments
+  for (std::list<PathSegment*>::iterator segment = pathSegments.begin(); segment != pathSegments.end(); segment++)
+  {
+    (*segment)->position = length;
+    std::list<double> localSwitchingPoints = (*segment)->getSwitchingPoints();
+    for (std::list<double>::const_iterator point = localSwitchingPoints.begin(); point != localSwitchingPoints.end();
+         point++)
+    {
+      switchingPoints.push_back(std::make_pair(length + *point, false));
+    }
+    length += (*segment)->getLength();
+    while (!switchingPoints.empty() && switchingPoints.back().first >= length)
+      switchingPoints.pop_back();
+    switchingPoints.push_back(std::make_pair(length, true));
+  }
+  switchingPoints.pop_back();
+}
+
+Path::Path(const Path& path) : length(path.length), switchingPoints(path.switchingPoints)
+{
+  for (std::list<PathSegment*>::const_iterator it = path.pathSegments.begin(); it != path.pathSegments.end(); it++)
+  {
+    pathSegments.push_back((*it)->clone());
+  }
+}
+
+Path::~Path()
+{
+  for (std::list<PathSegment*>::iterator it = pathSegments.begin(); it != pathSegments.end(); it++)
+  {
+    delete *it;
+  }
+}
+
+double Path::getLength() const
+{
+  return length;
+}
+
+PathSegment* Path::getPathSegment(double& s) const
+{
+  std::list<PathSegment*>::const_iterator it = pathSegments.begin();
+  std::list<PathSegment*>::const_iterator next = it;
+  next++;
+  while (next != pathSegments.end() && s >= (*next)->position)
+  {
+    it = next;
+    next++;
+  }
+  s -= (*it)->position;
+  return *it;
+}
+
+Eigen::VectorXd Path::getConfig(double s) const
+{
+  const PathSegment* pathSegment = getPathSegment(s);
+  return pathSegment->getConfig(s);
+}
+
+Eigen::VectorXd Path::getTangent(double s) const
+{
+  const PathSegment* pathSegment = getPathSegment(s);
+  return pathSegment->getTangent(s);
+}
+
+Eigen::VectorXd Path::getCurvature(double s) const
+{
+  const PathSegment* pathSegment = getPathSegment(s);
+  return pathSegment->getCurvature(s);
+}
+
+double Path::getNextSwitchingPoint(double s, bool& discontinuity) const
+{
+  std::list<std::pair<double, bool> >::const_iterator it = switchingPoints.begin();
+  while (it != switchingPoints.end() && it->first <= s)
+  {
+    it++;
+  }
+  if (it == switchingPoints.end())
+  {
+    discontinuity = true;
+    return length;
+  }
+  else
+  {
+    discontinuity = it->second;
+    return it->first;
+  }
+}
+
+std::list<std::pair<double, bool> > Path::getSwitchingPoints() const
+{
+  return switchingPoints;
+}
+
+const double Trajectory::eps = 0.000001;
+
+static double squared(double d)
+{
+  return d * d;
```

Comment:
is this function really worth adding another function call to stack? just a thought